### ISSUE TITLE
get(fun,'lval') and get(fun,'rval')

### ISSUE THREAD

**Issue by asgeirbirkis** [chebfun/chebfun#1608]
This is inspired by #1607 .

By using @aaustin141 's code and turning the profiler on:
```
Ntrials = 5;
t_total = 0;
profile on
chebfun(@(x) sign(x), 'trunc', 100);
for (k = 1:1:Ntrials)
        tic();
        chebfun(@(x) sign(x), 'trunc', 100);
        t = toc()
        t_total = t_total + t;
end
profile viewer
avg_time = t_total/Ntrials
```
reveals that the most self time is spent in `chebtech2.get`, in particular in the line `case fieldnames(f)` (the design of our `get` methods is material for its own issue, and it probably would be much more efficient to do a try-catch in the method -- fieldnames is an expensive operation).

Further investigating why we call `chebtech2.get` so often, it's evidently because of lines 30-35 in `chebfun.getValuesAtBreakpoints`:
```
    vals(1,:) = get(funs{1}, 'lval');
    % Take the mean of the FUNs on either side of the break:
    for k = 2:numFuns
        vals(k,:) = (get(funs{k-1}, 'rval') + get(funs{k}, 'lval'))/2;
    end
    vals(numFuns+1,:) = get(funs{numFuns}, 'rval');
```

But this is a bit crazy, because what happens at the bottom level in `chebtech2.get` is simply
```
    case 'lval'
        out = lval(f);
    case 'rval'
        out = rval(f);
```
So isn't it a bit weird not just to have `lval` and `rval` at the higher levels, and save all these `get` calls?

**Comment by nickhale** [comment]
This is weird. When I profile this (also on R2015b) I spend very little time in `get`. Most of my time is spent in `chebfun/chebcoeffs()` and in particular `chebfun.overlap()`. I can imprive this somewhat by changing lines 86-92 of `chebfun/chebcoeffs` to
```
    % Compute coefficients via inner products.
    d = f.domain;
    x = chebfun('x', d);
    if ( kind == 1 )
        w = 1./sqrt((x - d(1)).*(d(end) - x));
    elseif ( kind == 2 )
        w = sqrt((x - d(1)).*(d(end) - x));
    end
```
(Note that this only works as `chebfun/chebcoeffs' doesn't support quasimatrices anyway.)

Most of the time is now spent in multiplying the Chebyshev polynomials returned by `chebpoly()` with the Chebyshev weight function. This could probably be improved by vectorization, but the fact that `singfun` doesn't allow array-valued representation makes that tricky.

**Comment by asgeirbirkis** [comment]
@nickhale very weird indeed. Were you looking at total time or self time? In the original code snippet I pasted above, on development, I get 0.074 seconds in self time for chebfun.overlap (8.7s in total time), while chebtech2.get is 0.87 seconds in self time (1.24s in total time).

**Comment by nickhale** [comment]
Ah - you're right. I was simply looking at total time. 

In any case, I still think my suggestion is a good one. Does it change your timings at all?

**Comment by asgeirbirkis** [comment]
Yes, it does, bringing the time from around 3.05s to 2.3s (didn't time this extremely accurately, but definitely some speedup).

Brought it further down to 1.95s by replacing the line
```
x = chebfun('x', d);
```
from your snippet (the line is in the current version as well) to
```
x = chebfun([d(1); d(end)], d);
```
what timing results do you get with that?

**Comment by nickhale** [comment]
Changing to `x = chebfun([d(1); d(end)], d);` made things considerably worse for me.

I've pushed some ideas to `feature-lval-rval-speed`. Experiments suggest we should (a) remove calls to `fieldnames` when we can and (b) move them down the list of case options when we can't.

**Comment by asgeirbirkis** [comment]
> Changing to x = chebfun([d(1); d(end)], d); made things considerably worse for me.

I'm quite surprised by that, usually, the string construction is the slowest way to construct a chebfun. What do you get when you run the following:
```
d = [0 10];
tic, for i=1:400, x = chebfun('x', d); end, tString = toc;
tic, for i=1:400, x = chebfun(@(x) x, d); end, tAnon = toc;
tic, for i=1:400, x = chebfun([d(1); d(end)], d); end, tVec = toc;

fprintf('Time with string: %4.4f seconds\n', tString)
fprintf('Time with anon. function: %4.4f seconds\n', tAnon)
fprintf('Time with vector: %4.4f seconds\n', tVec)
```

My output is as follows:
```
Time with string: 3.3595 seconds
Time with anon. function: 2.2194 seconds
Time with vector: 1.9167 seconds
```

**Comment by nickhale** [comment]
Oh. My mistake. I misunderstood. Yeah, the should help too.

**Comment by asgeirbirkis** [comment]
> (Note that this only works as `chebfun/chebcoeffs' doesn't support quasimatrices anyway.)

So do we need to worry at all about quasimatrices in 103-113 of `chebfun/chebcoeffs`? Replacing
```
    f = mat2cell(f);
    T = chebpoly(0:(N-1), d, kind);
    T = mat2cell(T);
    for k = 1:N
        Tkw = T{k}.*w;
        for j = 1:numCols
            out(k,j) = innerProduct(f{j}, Tkw);
        end
    end
```
with the following lines (which are probably not yet dealing properly with array valued chebfuns):
```
T = chebpoly(0:(N-1), d, kind);
out = ( (f.*w)'*T )';
```
gives me on development.

Before this change (and with the changes above, but not with `lval` and `rval`:
```
>> tic, chebfun(@(x) sign(x), 'trunc', 100); toc
Elapsed time is 1.924937 seconds.
```
With this change:
```
>> tic, chebfun(@(x) sign(x), 'trunc', 100); toc
Elapsed time is 0.757758 seconds.
```

**Comment by nickhale** [comment]
Is there a branch / pull request related to this ticket?

**Comment by asgeirbirkis** [comment]
I'd been holding on doing anything performance related until finished reviewing #1584 , but with that sorted, I'd be happy to revisit this topic.

**Comment by nickhale** [comment]
I've just pushed an improvement. I'm now getting 
```
>> tic, chebfun(@(x) sign(x), 'trunc', 100); toc
Elapsed time is 0.220814 seconds.
```

**Comment by asgeirbirkis** [comment]
Can confirm significant speedup.

On development (using the `timeit` method suggested by @jrus):
```
>> f = @() chebfun(@(x) sign(x), 'trunc', 100);
>> timeit(f)
ans =
   1.478908886790500
```
on branch feature-lval-rval-speed:
```
>> f = @() chebfun(@(x) sign(x), 'trunc', 100);
>> timeit(f)
ans =
   0.240097057381000
```
I just pushed one line change replacing another chebtech2.get call with a call straight to the `vscale` method (which got called anyway inside the get).

Do you want to open a PR for this, or do you believe there are still further speedups to be gained?

**Comment by nickhale** [comment]
I'm sure there are more speed ups available, but I think we might as well merge this now.

I created a PR, but I noticed the `lval()` and `rval()` methods need documenting. I don't have time to do that now.

**Comment by asgeirbirkis** [comment]
I spotted a couple of further improvements (mostly getting rid of calls to feval) which I'll quickly push to be included .

**Comment by jrus** [comment]
So is `get` just incredibly slow as a general rule?

@nickhale is there anywhere else that `rval(some_fun)` is going to be called, besides `chebfun/getValuesAtBreakpoints`? If not, you could also just directly call `rval(funs{n}.onefun)`, etc. instead of adding a new method on funs.

**Comment by nickhale** [comment]
Yes, I think `get` should be avoided.

Unless it proves to make a massive difference in performance, I think `rval(some_fun)` is cleaner than `rval(some_fun.onefun)`, but I don't feel too strongly about it.

I think now that #1584 has been merged we can really start focussing on performance. However, I think that to do this we need to have a good idea of what problems are important and what we want to benchmark. We dabbled with this in v4 with `chebbench`, so that's probably a good place to start. I suggest we move this discussion to a new thread.

**Comment by jrus** [comment]
I think the key question for benchmarks / profiling is to figure out at a high level what kinds of problems people are trying to solve, and look at which bits take the most time in those. (On the other hand, what people try may be constrained by what’s currently acceptably fast.) I don’t really have a good sense of what common uses are like for Chebfun. Perhaps the place to start would be running some of the code from http://chebfun.org/examples/ with a profiler?

If you start a thread, I’m happy to toss in my ideas.

P.S. Have you folks seen this tool? http://undocumentedmatlab.com/blog/function-call-timeline-profiling

**Comment by asgeirbirkis** [comment]
Closed by #1625
